src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonArray.IList.cs

```diff
@@ -114,6 +116,116 @@ public void RemoveAt(int index)
+            if (freeIndex >= size) return 0;
+
+            int current = freeIndex + 1;
+            while (current < size)
+            {
+                // Find the first item which needs to be kept.
+                while (current < size && match(list[current])) current++;
+
+                if (current < size)
+                {
+                    DetachParent(list[freeIndex]);
+
+                    // copy item to the free slot.
+                    list[freeIndex++] = list[current];
+                    list[current] = null; // prevent it from being detached from the parent in the following for-loop
+                    ++current;
+                }
+            }
+
+            // Detach trailing nodes, as they are skipped in the loop above
+            int removedLength = size - freeIndex;
+            for (int i = 0; i < removedLength; i++)
+            {
+                DetachParent(list[freeIndex + i]);
+            }
+
+            list.RemoveRange(freeIndex, removedLength);
+            return removedLength;
+        }
+
+        /// <summary>
+        ///   Removes a range of elements from the <see cref="JsonArray"/>.
+        /// </summary>
+        /// <param name="index">The zero-based starting index of the range of elements to remove.</param>
+        /// <param name="count">The number of elements to remove.</param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        ///   <paramref name="index"/> or <paramref name="count"/> is less than 0.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        ///   <paramref name="index"/> and <paramref name="count"/> do not denote a valid range of elements in the <see cref="JsonArray"/>.
+        /// </exception>
+        public void RemoveRange(int index, int count)
+        {
+            if (index < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException_NeedNonNegNum(nameof(index));
+            }
+
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException_NeedNonNegNum(nameof(count));
+            }
+
+            if (_list is not { } list)
+            {
+                if (index == 0 && count == 0)
+                {
+                    return;
+                }
+                else
+                {
+                    ThrowHelper.ThrowArgumentException_InvalidOffLen();
+                    return;
+                }
+            }
+
+            if (list.Count - index < count)
+            {
+                ThrowHelper.ThrowArgumentException_InvalidOffLen();
+            }
+
+            if (count > 0)
+            {
+                for (int i = 0; i < count; i++)
+                {
+                    DetachParent(list[index + i]);
+                    // There's no need to assign nulls because List<>.RemoveRange calls
+                    // Array.Clear on the removed partition.
+                }
+
+                list.RemoveRange(index, count);
```

Comment:
Is there a possibility for this `RemoveRange` call to throw here given the argument validation you've already done?

Reply:
> Is there a possibility for this `RemoveRange` call to throw here given the argument validation you've already done?

I don't think so. I adapted the code from `List<T>.RemoveRange` so they have the same error-handling logic currently. Besides, `JsonNode.set_Parent()` cannot be overidden which shouldn't enable the `JsonArray` to be altered midway.